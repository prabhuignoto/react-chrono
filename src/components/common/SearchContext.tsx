import React, { createContext, useState, useContext } from 'react';

interface SearchContextType {
  searchTerm: string;
  setSearchTerm: (term: string) => void;
}

// Create a context with default values
const SearchContext = createContext<SearchContextType>({
  searchTerm: '',
  setSearchTerm: () => {},
});

/**
 * Provider component for the search context
 */
export const SearchProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const [searchTerm, setSearchTerm] = useState<string>('');

  return (
    <SearchContext.Provider value={{ searchTerm, setSearchTerm }}>
      {children}
    </SearchContext.Provider>
  );
};

/**
 * Custom hook to use the search context
 */
export const useSearch = (): SearchContextType => {
  const context = useContext(SearchContext);

  if (context === undefined) {
    throw new Error('useSearch must be used within a SearchProvider');
  }

  return context;
};

export default SearchContext;
