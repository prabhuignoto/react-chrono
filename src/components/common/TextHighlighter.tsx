import React from 'react';
import styled from 'styled-components';
import { Theme } from '@models/Theme';

interface HighlightedTextProps {
  text: string;
  searchTerm: string;
  theme: Theme;
}

// Styled component for highlighted text
const HighlightSpan = styled.span<{ theme: Theme }>`
  background-color: ${({ theme }) =>
    theme.primary}33; /* Primary color with 20% opacity */
  color: ${({ theme }) => theme.primary};
  font-weight: 500;
  border-radius: 2px;
  padding: 0 2px;
`;

/**
 * Component that highlights occurrences of a search term within text
 */
const TextHighlighter: React.FC<HighlightedTextProps> = ({
  text,
  searchTerm,
  theme,
}) => {
  // If no search term or no text, just return the original text
  if (!searchTerm || !text) {
    return <>{text}</>;
  }

  // Escape special regex characters in the search term
  const escapedSearchTerm = searchTerm.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

  // Create a case-insensitive regular expression to find all occurrences
  const regex = new RegExp(`(${escapedSearchTerm})`, 'gi');

  // Split the text by the search term
  const parts = text.split(regex);

  // Return the text with highlighted parts
  return (
    <>
      {parts.map((part, i) =>
        regex.test(part) ? (
          <HighlightSpan key={i} theme={theme}>
            {part}
          </HighlightSpan>
        ) : (
          <React.Fragment key={i}>{part}</React.Fragment>
        ),
      )}
    </>
  );
};

export default TextHighlighter;
