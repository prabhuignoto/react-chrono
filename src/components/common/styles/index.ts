import { Theme } from '@models/Theme';
import { css } from 'styled-components';

export const ScrollBar = css<{ theme: Theme }>`
  scrollbar-color: ${(p) => p.theme?.primary} default;
  scrollbar-width: thin;

  &::-webkit-scrollbar {
    width: 0.3em;
  }

  &::-webkit-scrollbar-track {
    box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.2);
  }

  &::-webkit-scrollbar-thumb {
    background-color: ${(p) => p.theme?.primary};
    outline: 1px solid ${(p) => p.theme?.primary};
  }
`;
