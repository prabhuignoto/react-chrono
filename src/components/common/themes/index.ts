import { Theme } from '@models/Theme';

export const defaultTheme: Theme = {
  cardBgColor: '#ffffff',
  cardDetailsBackGround: '#ffffff',
  cardDetailsColor: '#000',
  cardMediaBgColor: '#f5f5f5',
  cardSubtitleColor: '#000',
  cardTitleColor: '#007FFF',
  detailsColor: '#000',
  iconBackgroundColor: '#007FFF',
  nestedCardBgColor: '#f5f5f5',
  nestedCardDetailsBackGround: '#f5f5f5',
  nestedCardDetailsColor: '#000',
  nestedCardSubtitleColor: '#000',
  nestedCardTitleColor: '#000',
  primary: '#007FFF',
  secondary: '#ffdf00',
  titleColor: '#007FFF',
  titleColorActive: '#007FFF',
  toolbarBgColor: '#f5f5f5',
  toolbarBtnBgColor: '#fff',
  toolbarTextColor: '#000',
};

export const darkTheme: Theme = {
  cardBgColor: '#191919',
  cardDetailsBackGround: '#191919',
  cardDetailsColor: '#ffff0f',
  cardMediaBgColor: '#2f2f2f',
  cardSubtitleColor: '#ffffff',
  cardTitleColor: '#007FFF',
  detailsColor: '#ffffff',
  iconBackgroundColor: '#007FFF',
  nestedCardBgColor: '#333333',
  nestedCardDetailsBackGround: '#333333',
  nestedCardDetailsColor: '#ffffff',
  nestedCardSubtitleColor: '#ffffff',
  nestedCardTitleColor: '#ffffff',
  primary: '#007FFF',
  secondary: '#ffdf00',
  titleColor: '#007FFF',
  titleColorActive: '#007FFF',
  toolbarBgColor: '#555',
  toolbarBtnBgColor: '#222',
  toolbarTextColor: '#ffffff',
};
