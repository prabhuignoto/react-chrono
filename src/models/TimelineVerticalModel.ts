import { Theme } from './Theme';
import { Scroll } from './TimelineHorizontalModel';
import { TimelineCardModel } from './TimelineItemModel';
import { Media } from './TimelineMediaModel';
import { TimelineProps } from './TimelineModel';

export type Props = Pick<
  TimelineProps,
  | 'flipLayout'
  | 'theme'
  | 'mode'
  | 'timelineCircleDimension'
  | 'lineWidth'
  | 'cardHeight'
  | 'enableOutline'
  | 'disableClickOnCircle'
  | 'cardLess'
> & {
  alternateCards?: boolean;
  hasFocus?: boolean;
  onClick: (id?: string) => void;
  onElapsed?: (id?: string) => void;
  slideItemDuration?: number;
  slideShowRunning?: boolean;
  theme?: Theme;
};

type VerticalModel = Pick<
  Props,
  | 'alternateCards'
  | 'hasFocus'
  | 'onClick'
  | 'onElapsed'
  | 'slideShowRunning'
  | 'mode'
  | 'timelineCircleDimension'
  | 'lineWidth'
  | 'disableClickOnCircle'
  | 'cardLess'
> & {
  active?: boolean;
  className: string;
  id?: string;
};

export interface VerticalCircleModel extends VerticalModel {
  iconChild?: React.ReactNode;
  onActive: (pointOffset: number) => void;
}

export interface VerticalItemModel extends VerticalModel {
  cardDetailedText?: string | string[];
  cardSubtitle?: string;
  cardTitle?: string;
  contentDetailsChildren?: React.ReactNode;
  iconChild?: React.ReactNode;
  index: number;
  media?: Media;
  onActive: (
    pointOffset: number,
    contentHeight: number,
    contentOffset: number,
  ) => void;
  onShowMore?: () => void;
  title?: string;
  url?: string;
  visible?: boolean;
}

export type TimelineVerticalModel = Pick<
  Props,
  | 'alternateCards'
  | 'enableOutline'
  | 'mode'
  | 'onClick'
  | 'onElapsed'
  | 'slideShowRunning'
  | 'theme'
  | 'hasFocus'
  | 'cardLess'
> & {
  activeTimelineItem: number;
  autoScroll: (s: Partial<Scroll>) => void;
  childrenNode?: React.ReactNode;
  contentDetailsChildren?: React.ReactNode;
  iconChildren?: React.ReactNode;
  items: TimelineCardModel[];
  onOutlineSelection?: (index: number) => void;
};
