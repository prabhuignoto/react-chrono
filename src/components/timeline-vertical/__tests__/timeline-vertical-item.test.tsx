import { VerticalItemModel } from '@models/TimelineVerticalModel';
import { describe, expect, it } from 'vitest';
import { customRender } from '../../common/test';
import { providerProps } from '../../common/test/index';
import TimelineVerticalItem from '../timeline-vertical-item';

// Mock ResizeObserver
global.ResizeObserver = class ResizeObserver {
  observe() {}
  unobserve() {}
  disconnect() {}
};

const commonProps: VerticalItemModel = {
  // complete the rest of the properties
  active: false,
  alternateCards: false,
  cardDetailedText: '',
  cardSubtitle: '',
  cardTitle: '',
  className: '',
  contentDetailsChildren: null,
  hasFocus: false,
  iconChild: null,
  id: '',
  index: 1,
  media: {
    source: {
      type: 'IMAGE',
      url: 'https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_272x92dp.png',
    },
    type: 'IMAGE',
  },
  onActive: () => {},
  onClick: () => {},
  onElapsed: () => {},
  slideShowRunning: false,
  timelineContent: null,
  title: 'vertical item title',
  url: '',
  visible: false,
};

describe('Timeline vertical item', () => {
  it('Should match snapshot', () => {
    const { container } = customRender(
      <TimelineVerticalItem {...commonProps} />,
      { providerProps },
    );

    expect(container).toMatchSnapshot();
  });

  //should render the title

  it('Should render the title', () => {
    const { getByText } = customRender(
      <TimelineVerticalItem {...commonProps} />,
      { providerProps },
    );

    expect(getByText('vertical item title')).toBeInTheDocument();
  });
});
