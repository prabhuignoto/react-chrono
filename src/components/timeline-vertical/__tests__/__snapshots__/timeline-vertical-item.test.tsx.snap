// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Timeline vertical item > Should match snapshot 1`] = `
<div>
  <div
    class="sc-dnwKUv jBiQiW  vertical-item-row"
    data-testid="vertical-item-row"
    role="listitem"
  >
    <div
      class="sc-fLQRDB cpMBSd"
      mode="VERTICAL_ALTERNATING"
    >
      <div
        class="sc-iAEyYk ehSAtw timeline-item-title title"
        font-size="1.5rem"
      >
        vertical item title
      </div>
    </div>
    <div
      class="sc-kgKVFY bYjqVL card-content-wrapper "
    >
      <section
        class="sc-jegxcv bXREYn timeline-card-content  card"
        mode="VERTICAL_ALTERNATING"
        tabindex="0"
      >
        <div
          class="sc-eACynP jTJJiI card-media-wrapper card-media"
          mode="VERTICAL_ALTERNATING"
        >
          <div
            class="sc-cjERFW hasYaA"
            height="200"
          >
            <img
              class="sc-bZPPFW dwXfbd"
              loading="lazy"
              mode="VERTICAL_ALTERNATING"
              role="img"
              src="https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_272x92dp.png"
            />
          </div>
        </div>
        <div
          aria-expanded="false"
          class="sc-dQelHR exXkiz show-less card-description card-text"
        />
        <span
          class="sc-kiLEMZ ehIRXU"
          dir=""
          offset="-8"
        />
      </section>
    </div>
    <div
      class="sc-kZGvTt iiUlHY"
      data-testid="tree-leaf"
      role="button"
    >
      <div
        aria-label="select timeline"
        class="sc-iMfspA dptdC  timeline-vertical-circle"
        data-testid="tree-leaf-click"
        role="button"
      >
        <div
          class="sc-kFuwaP naOQz"
        />
      </div>
    </div>
  </div>
</div>
`;
